src/experimental/tools/videoThumbnailLoader/init_source_buffer.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  EMPTY,
+  Observable,
+  of as observableOf,
+  ReplaySubject,
+  Subscription,
+  merge as observableMerge,
+  Subject,
+} from "rxjs";
+import {
+  filter,
+  map,
+  mapTo,
+  mergeMap,
+  tap,
+} from "rxjs/operators";
+import {
+  ISegmentLoaderContent,
+  ISegmentLoaderEvent,
+} from "../../../core/fetchers/segment/create_segment_loader";
+import { AudioVideoSegmentBuffer } from "../../../core/segment_buffers/implementations";
+import { ITransportAudioVideoSegmentParser } from "../../../transports";
+import prepareSourceBuffer from "./prepare_source_buffer";
+import { IContentInfos } from "./types";
+
+let mediaSourceSubscription: Subscription | undefined;
+let currentContentInfos: IContentInfos | undefined;
+
+let sourceBuffer$: ReplaySubject<AudioVideoSegmentBuffer> =
+  new ReplaySubject();
+
+/**
+ * Get current source buffer :
+ * - If already created for media element, reuse
+ * - Else, create a new one
+ * @param {Object} contentInfos
+ * @param {HTMLVideoElement} element
+ * @returns {Observable}
+ */
+export function initSourceBuffer$(
+  contentInfos: IContentInfos,
+  element: HTMLVideoElement,
+  { segmentLoader,
+    segmentParser }: {
+    segmentLoader: (
+      x : ISegmentLoaderContent
+    ) => Observable<ISegmentLoaderEvent<Uint8Array |
+                                        ArrayBuffer |
+                                        null>>;
+    segmentParser: ITransportAudioVideoSegmentParser;
+  }
+): Observable<AudioVideoSegmentBuffer> {
+  const mediaSourceError$ = new Subject();
+  if (mediaSourceSubscription === undefined) {
+    mediaSourceSubscription =
```

Comment:
That means that we will rely on the same SourceBuffer even when the codec is not the same (e.g. avc to hevc) - despite the fact that many browsers are not able to do that trick

Isn't that problematic? We could postpone that very rare feature for a new version though.

Reply:
It is for now a rare case, and we don't really know what platform has issues with changing codecs.
But we could consider doing it later (reloading mediasource).

Reply:
Yes, but we could risk subtle future race conditions as this process here is asynchronous and never cancelled (we could be creating SourceBuffers on since inactive MediaSource?).

Again that's not the case for now so no problem but it's still something to keep in mind